### ISSUE TITLE
Feature request: Module APIs

### ISSUE THREAD

**Issue by Macil** [justjake/quickjs-emscripten#55]
It would be great if this project supported es2015 modules and dynamic `import()`. I believe QuickJS supports this if its function `JS_SetModuleLoaderFunc` is called with an appropriate callback that returned the contents for a give module URL.

My idea for an API would look like this:

```ts
import { getQuickJS } from 'quickjs-emscripten'

async function main() {
  const QuickJS = await getQuickJS()
  const vm = QuickJS.createVm()

  vm.setModuleLoaderFunc(name => {
    switch (name) {
      case '/dir/a.js':
        return `
          export let x = 'b has not run'
          export function setX(_x) {
            x = _x
          }
        `
      case '/dir/b.js':
        return `
          import {setX} from './a.js'
          setX('b has run')
        `
    }
    // Returning null/undefined in this function causes the module loading to fail with a generic error.
    // Alternatively an error can be thrown here and its message is passed into QuickJS.
  })

  // mod here is a QuickJSHandle representing the kind of object you get from the
  // statement `import * as mod from '...'`.
  const mod = vm.runModule(`
    import {x} from './dir/a.js'
    import './dir/b.js'
    export {x}
    export function test() { return 123; }
  `)

  // should output "mod x: b has run"
  console.log('mod x:', vm.getProp(mod, 'x').consume(x => vm.dump(x)));

  const testHandle = vm.getProp(mod, 'test')
  const result = vm.callFunction(testHandle, vm.undefined)
  if (result.error) {
    console.log('Execution failed:', vm.dump(result.error))
    result.error.dispose()
  } else {
    console.log('Success:', vm.dump(result.value))
    result.value.dispose()
  }
  // should output "Success: 123"

  testHandle.dispose()
  mod.dispose()
  vm.dispose()
}

main()
```

This introduces the `vm.setModuleLoaderFunc` and `vm.runModule` functions.

`vm.runModule` executes code in module mode instead of script mode so that static imports work, and it returns a reference to the kind of object you get from `import * as ... from '...'` statements so that the module's exports can be accessed easily.

`vm.setModuleLoaderFunc` is meant to be a straight-forward wrapper around QuickJS's `JS_SetModuleLoaderFunc`, and can be replaced at any time, can dynamically generate modules, can do side-effects like logging, etc. (Ideally it would support an asynchronous callback so it could fetch remote modules on demand, but sadly it seems like QuickJS only supports synchronous module loader functions. I'm interested in trying to address that in the future, but that seems like it would take changes in QuickJS itself and I would find the existing sync API still useful.)

Note that the function passed to `JS_SetModuleLoaderFunc` and therefore also the function passed here to `vm.setModuleLoaderFunc` receives the resolved name of the module, not necessarily the literal value in the `import ... from '...path...'` statement. I believe QuickJS handles this part automatically, but if this behavior needs to be customized then a resolver function can also be supplied to QuickJS through `JS_SetModuleLoaderFunc`, and maybe this project could expose an API for passing a custom function there too. (Maybe `vm.setModuleLoaderFunc` would have a second optional parameter, or an overload where you give it a `{moduleLoader: Function, resolver?: Function}` object.)

For the common case where `vm.setModuleLoaderFunc` is called with a function that just maps names to fixed values, there could be a helper function like this:

```ts
import { getQuickJS, createFixedModuleLoader } from 'quickjs-emscripten'
// ...
  vm.setModuleLoaderFunc(createFixedModuleLoader({
    '/dir/a.js': `
      export let x = 'b has not run'
      export function setX(_x) {
        x = _x
      }
    `,
    '/dir/b.js': `
      import {setX} from './a.js'
      setX('b has run')
    `
  }))
```

Ideally `vm.setModuleLoaderFunc` being used would allow dynamic `import()` to work within code executed with either `vm.runModule` or `vm.evalCode`.

**Comment by justjake** [comment]
Yeah! I’d really like to enable this. I think the proposed API makes sense, but doesn’t handle the case of defining a module implementation itself using the API. For that we’d need some APIs to create module definition objects too. 

I also have a use-case for modules, but I need to be able to load module contents asynchronously. Unfortunately QuickJS doesn’t support an async loading API, so I’ve been working on a new build strategy using Emscripten ASYNCIFY in this PR: #54 

My plan is to introduce modules API there. That change also includes new APIs for Runtime and Context, which are conflated by the current QuickJSVm class we offer. The work was blocked by a bug I reported in Emscripten until recently. Now the bug is fixed upstream, but I’m not sure if there’s been a release that contains the fix.

**Comment by justjake** [comment]
Synchronous string module loading is now passing a test in #54. There’s still some work left on async module loading.

**Comment by Macil** [comment]
This sounds great!

One downside of relying on asyncify to accomplish async module loading is that fetching multiple modules in parallel won't be possible. As far as I can tell, that would require support from QuickJS. I've sent a message to the mailing list about that at https://www.freelists.org/post/quickjs-devel/Ability-to-asynchronously-load-modules-allowing-parallel-fetching.

**Comment by justjake** [comment]
@Macil what use case do you have for quickjs-emscripten where you need to async load modules in parallel? For static imports at the top of the module, the paths can be extracted with a simple AST pass for preloading, using any off the shelf JS parsing library. It’s quite quick and easy to do with ESQuery/Esprima or TSQuery/Typescript.

**Comment by Macil** [comment]
Partly I was aiming at making something Deno-compatible including working fully with dynamic imports. But dynamic imports aren't something I strictly need now so for now for my project I'm going to go on a route like that.

My project has users send code to a server where it runs continuously in a sandbox. I think I'll either have the users send the result of `deno bundle` (which bundles a project into a single .js file which would also happen to remove the need for anything to do with module APIs here), or send a zip of modules and a json file describing which imports where refer to which modules.

**Comment by justjake** [comment]
Even with ASYNCIFY this turns out to be a problem. Asyncify can only suspend one promise at a time, but the natural recursive flow of imports violates this restriction:

- eval imports  module A
- ASYNCIFY suspend to async load module A
- module A needs to be compiled
- module A requests async suspend to load module B
- Crash

**Comment by justjake** [comment]
Closed by #54 - please check out the beta 😄.

I plan to merge #59 which changes some minor private APIs and type imports before I cut the next release.
